pkg/wireguard/agent/agent.go

```diff
@@ -98,7 +98,13 @@ func NewAgent(privKeyPath string, localNodeStore *node.LocalNodeStore) (*Agent,
 	optOut := false
 	localNodeStore.Update(func(localNode *node.LocalNode) {
 		optOut = localNode.OptOutNodeEncryption
-		localNode.EncryptionKey = types.StaticEncryptKey
+		if !optOut {
+			// Always propagate a node encryption key, unless the local
+			// node opted out of node-to-node encryption. If node-to-node
+			// encryption is disabled, the EncryptionKey field is ignored by
+			// remote nodes (see node/manager.nodeAddressHasEncryptKey).
+			localNode.EncryptionKey = types.StaticEncryptKey
+		}
```

Comment:
Now that I'm going through this flow once more, I'm pretty confident that this would not work as expected. Indeed, the `localNode.OptOutNodeEncryption` value is initialized by the [initFromK8s](https://github.com/cilium/cilium/blob/355f2ffef29ad30fbdfcbf3ff20b5356e7bc994e/daemon/cmd/local_node_sync.go#L194) function (indirectly) invoked by the [LocalNodeStore start hook](https://github.com/cilium/cilium/blob/355f2ffef29ad30fbdfcbf3ff20b5356e7bc994e/pkg/node/local_node_store.go#L100). But here we are in the context of a provided function, which is executed earlier than start hooks (and thus `optOut` will always be false). 

One solution could be initializing the encryption key as part of `initFromK8s` as well, although that would still leave open how to correctly retrieve the `OptOutNodeEncryption` value for the wireguard module (AFAICT, it seems to be used for status reporting only). Setting the encryption key in a start hook here instead would be prone to race conditions, as observers (e.g., the nodediscovery package) may first see and propagate the node object without the encryption key set.

Reply:
Thanks for looking into this. I put the PR back into draft because local testing also revealed that this mechanism was not working.